File: app/src/main/java/com/example/newsfeed/NewsApp/UI/NewsViewModel.java
package com.example.newsfeed.NewsApp.UI;

import android.app.ProgressDialog;
import android.util.Log;
import android.widget.ProgressBar;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.newsfeed.NewsApp.NewsActivity;
import com.example.newsfeed.NewsApp.Repository.NewsRepository;
import com.example.newsfeed.NewsApp.api.NewsApi;
import com.example.newsfeed.NewsApp.api.RetrofitInstance;
import com.example.newsfeed.NewsApp.models.ArticlesItem;
import com.example.newsfeed.NewsApp.models.NewsResponse;
import com.example.newsfeed.NewsApp.util.Constants;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class
NewsViewModel extends ViewModel {
    public NewsRepository newsRepository;

    public NewsViewModel(NewsRepository newsRepository) {
        this.newsRepository = newsRepository;
    }

    public NewsViewModel() {
        // zero-args constructor
    }

    private MutableLiveData<List<ArticlesItem>> newsArticlesLiveData;
    private MutableLiveData<List<ArticlesItem>> searchNewsLiveData;

    private static final String TAG = "NewsViewModel.Articles";
    public static final String TAG1 = "Articles";
    private String Message;
private String searchingText;
    private void initNews(String message) {
        NewsApi newsApi = RetrofitInstance.newsApi;
        Call<NewsResponse> call = newsApi.getBreakingNews("in", message, Constants.API_KEY);
        call.enqueue(new Callback<NewsResponse>() {
            @Override
            public void onResponse(Call<NewsResponse> call, Response<NewsResponse> response) {
                if (response.isSuccessful() && response.body().getArticles() != null) {
                    List<ArticlesItem> articlesItems = response.body().getArticles();
                    newsArticlesLiveData.postValue(articlesItems);
                } else {
                    Log.i(TAG, "onResponse but articles not successful");
                }
            }
            @Override
            public void onFailure(Call<NewsResponse> call, Throwable t) {
                Log.e(TAG, "onFailure : " + t.getLocalizedMessage());
            }
        });
    }
    // liveData method
    public LiveData<List<ArticlesItem>> getBreakingNews(String message) {

        if (newsArticlesLiveData == null) {
            newsArticlesLiveData = new MutableLiveData<>();
            this.Message = message;
            Log.i("LiveData", "getBreakingNews " + message);
            // in the below function we pass the articles inside our liveData that we have
            // fetched from NewsApi
            initNews(message);
        } else if (!message.equals(Message)) {
            newsArticlesLiveData = new MutableLiveData<>();
            this.Message = message;
            initNews(message);
        }
        return newsArticlesLiveData;
    }
    public LiveData<List<ArticlesItem>> SearchNews(String searchQuery, ProgressDialog progressDialog) {
        if (searchNewsLiveData == null) {
            searchNewsLiveData = new MutableLiveData<>();
            searchingText=searchQuery;
            searchNewsFromApi(searchQuery,progressDialog);
        }
        else if(!searchQuery.equalsIgnoreCase(searchingText))
        {
            searchNewsLiveData=new MutableLiveData<>();
            searchingText=searchQuery;
            searchNewsFromApi(searchQuery,progressDialog);
        }else {
            progressDialog.dismiss();
        }return searchNewsLiveData;
    }
    private void searchNewsFromApi(String  searchQuery, ProgressDialog progressDialog) {
        NewsApi newsApi=RetrofitInstance.newsApi;
        int searchNewsPage = 1;
        Call<NewsResponse> call = newsApi.SearchNews(searchQuery, searchNewsPage,Constants.API_KEY);
        call.enqueue(new Callback<NewsResponse>() {
            @Override
            public void onResponse(Call<NewsResponse> call, Response<NewsResponse> response) {
                Log.e(TAG, "onResponse  : " + response.body().getTotalResults());
                if (response.isSuccessful() && response.body().getArticles() != null) {
                    List<ArticlesItem> articlesItems=response.body().getArticles();
                    searchNewsLiveData.setValue(articlesItems);
                    Log.e(TAG1, "articles from searchNews : " + searchNewsLiveData.getValue().toString());
                } else {
                    Log.i(TAG, "onResponse in SearchNews but articles not successful");
                }
                progressDialog.dismiss();
            }

            @Override
            public void onFailure(Call<NewsResponse> call, Throwable t) {
                Log.e(TAG, "onFailure : " + t.getLocalizedMessage());
                progressDialog.setCancelable(true);
            }
        });
    }


}

File: app/src/main/java/com/example/newsfeed/NewsApp/UI/NewsViewModel.kt
package com.example.newsfeed.NewsApp.UI

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.net.ConnectivityManager.*
import android.net.NetworkCapabilities.*
import android.os.Build
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.example.newsfeed.NewsApp.NewsApplication
import com.example.newsfeed.NewsApp.Repository.NewsRepository
import com.example.newsfeed.NewsApp.models.ArticlesItem
import com.example.newsfeed.NewsApp.models.NewsResponse
import com.example.newsfeed.NewsApp.util.Resource
import kotlinx.coroutines.launch
import retrofit2.Response
import java.io.IOException


class NewsViewModel(app: Application, val newsRepository: NewsRepository) : AndroidViewModel(app) {

    val categoryNews: MutableLiveData<Resource<NewsResponse>> = MutableLiveData()
    var categoryNewPageNumber = 1
    var categoryNewsResponse: NewsResponse? = null


    val searchNews: MutableLiveData<Resource<NewsResponse>> = MutableLiveData()
    var searchNewsPageNumber = 1
    var message: String? = null
    var searchNewsResponse: NewsResponse? = null

    var newCategory:String?=null
    var oldCategory:String?=null
    var newSearchQuery:String? = null
    var oldSearchQuery:String? = null

    fun getCategoryNews(countryCode: String, category: String) = viewModelScope.launch {
      //  categoryNews.postValue(Resource.Loading())
safeCategoryNewsCall(countryCode, category)
    }

    fun getSearchNews(searchQuery: String) = viewModelScope.launch {
      //  Log.d("SearchText","query in getSearchNews() $searchQuery")

safeSearchNewsCall(searchQuery)
    }

    // database functionality

    fun savedArticle(articlesItem: ArticlesItem) = viewModelScope.launch {
        newsRepository.upsert(articlesItem)
    }

    fun getSavedNewsToDisplay() = newsRepository.getSavedArticles()

    fun deleteArticles(articlesItem: ArticlesItem) = viewModelScope.launch {
        newsRepository.deleteArticles(articlesItem)
    }


    private fun getCategoryNewsResponse(response: Response<NewsResponse>): Resource<NewsResponse> {
        if (response.isSuccessful) {
            response.body()?.let { resultResponse ->
                categoryNewPageNumber++
                if (categoryNewsResponse == null|| oldCategory!=newCategory) {
                    categoryNewPageNumber=1
                    oldCategory=newCategory
                    categoryNewsResponse = resultResponse
                } else {
                    categoryNewPageNumber++
                    val oldArticlesItem = categoryNewsResponse?.articles
                    val newArticlesItem = resultResponse.articles
                    oldArticlesItem?.addAll(newArticlesItem)
                }
                return Resource.Success(categoryNewsResponse ?: resultResponse)
            }
        }
        return Resource.Error(response.message())
    }

    private fun getSearchNewsResponse(response: Response<NewsResponse>): Resource<NewsResponse> {
        if (response.isSuccessful) {
            response.body()?.let {resultResponse->
                searchNewsPageNumber++
                if(searchNewsResponse == null || newSearchQuery != oldSearchQuery) {
                    searchNewsPageNumber = 1
                    oldSearchQuery = newSearchQuery
                    searchNewsResponse = resultResponse
                } else {
                    searchNewsPageNumber++
                    val oldArticles = searchNewsResponse?.articles
                    val newArticles = resultResponse.articles
                    oldArticles?.addAll(newArticles)
                }

                return Resource.Success (searchNewsResponse?:resultResponse)
            }
        }
        return Resource.Error(response.message())
    }

    fun sendingCategory(msg: String) {
        message = msg
        message?.let { getCategoryNews("in", category = it) }
    }

    private suspend fun safeCategoryNewsCall(countryCode: String,category: String){
        newCategory=category
        categoryNews.postValue(Resource.Loading())
        try {
            if(hasInternetConnection()) {
                val response = newsRepository.getCategoryNews(countryCode, category, categoryNewPageNumber)
                categoryNews.postValue(getCategoryNewsResponse(response))
            }
            else{
categoryNews.postValue(Resource.Error("No Internet Connection"))
            }
            }
        catch (t:Throwable){
when(t){
    is IOException->categoryNews.postValue(Resource.Error("Network Failure"))
    else ->categoryNews.postValue(Resource.Error("Conversion Error"))
}
        }
    }
    private suspend fun safeSearchNewsCall(searchQuery: String){
        newSearchQuery=searchQuery
        searchNews.postValue(Resource.Loading())
        try {
            if(hasInternetConnection()) {
                Log.d("SearchText","safeSearchNewsCall inside try inside if $searchQuery")
                val response = newsRepository.searchNews(searchQuery,searchNewsPageNumber)
                searchNews.postValue(getSearchNewsResponse(response))
            }
            else{
                Log.d("SearchText","safeSearchNewsCall inside try inside else $searchQuery")

                searchNews.postValue(Resource.Error("No Internet Connection"))
            }
        }
        catch (t:Throwable){
            when(t){
                is IOException->searchNews.postValue(Resource.Error("Network Failure"))
                else ->searchNews.postValue(Resource.Error("Conversion Error"))
            }
        }
    }


    private fun hasInternetConnection(): Boolean {
        val connectivityManager = getApplication<NewsApplication>().getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val activeNetwork = connectivityManager.activeNetwork ?: return false
            val capabilities = connectivityManager.getNetworkCapabilities(activeNetwork)
                    ?: return false
            return when {
                capabilities.hasTransport(TRANSPORT_WIFI) -> true
                capabilities.hasTransport(TRANSPORT_CELLULAR) -> true
                capabilities.hasTransport(TRANSPORT_ETHERNET) -> true
                else -> false
            }
        } else {
            connectivityManager.activeNetworkInfo?.run {
                return when (type) {
                    TYPE_WIFI -> true
                    TYPE_MOBILE -> true
                    TYPE_ETHERNET -> true
                    else -> false
                }
            }
        }
        return false
    }
}